src/savecontext.cc

```diff
@@ -0,0 +1,98 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0. This product includes software
+// developed at Datadog (https://www.datadoghq.com/). Copyright 2021-Present
+// Datadog, Inc.
+
+#include "savecontext.hpp"
+
+#include <algorithm>
+#include <cstring>
+#include <pthread.h>
+
+#define XSTR(s) STR(s)
+#define STR(s) #s
+
+#ifdef __x86_64__
+// hardcode register index in mask
+// Did not find a way to reuse enum values from perf-archmap.h
+#  define RBX_IDX 1
+#  define RBP_IDX 6
+#  define RSP_IDX 7
+#  define RIP_IDX 8
+#  define R12_IDX 16
+#  define R13_IDX 17
+#  define R14_IDX 18
+#  define R15_IDX 19
+
+#  define DEST_MEM(reg) XSTR(reg##_IDX) "*8(%rdi)\n"
+#  define SAVE_REG(reg) "movq %" #  reg ", " DEST_MEM(reg)
+
+static __attribute__((noinline, naked)) void
+save_registers(ddprof::span<uint64_t, PERF_REGS_COUNT>) {
+  // The goal here is to capture the state of registers after the return of this
+  // function. That is why this function must not be inlined.
+
+  // clang-format off
+    asm (
+    // Retrieve caller save registers
+    // Callee save register are not needed since they could contain anything after function return
+    SAVE_REG(RBX)
+    SAVE_REG(RBP)    
+    SAVE_REG(R12)
+    SAVE_REG(R13)
+    SAVE_REG(R14)
+    SAVE_REG(R15)
+    // Bump the stack by 8 bytes to remove the return address, 
```

Comment:
thanks for the comments !

Reply:
Only works for x86_64 for now, I will do the implementation for ARM later.